### PR TITLE
#2686 Breach column

### LINKED ISSUE #2686: vaccine breach column
## Is your feature request related to a problem? Please describe.

Need to add a vaccine breach column to the customer invoice and stocktake batch pages

## Describe the solution you'd like

A column with a breach icon when the underlying batch is in breach which will open the breach modal

## Implementation

N/A

## Describe alternatives you've considered

N/A

## Additional context

N/A

### PR DESCRIPTION
**Pull request by joshxg**
Fixes #2686 

## Change summary

- Adds a breach column to `StocktakeBatchModal` and `CustomerInvoice` page.
- Changed up the `BreachActions` - think they're nicer now!

breach1

## Testing

N/A

### Related areas to think about

-

### CODE REVIEW

File: src/database/DataTypes/ItemBatch.js [hunk 1]
```diff
@@ -30,6 +30,10 @@ export class ItemBatch extends Realm.Object {
     return this.location?.isInBreach() ?? false;
   }
 
+  get breaches() {
```
Comment (joshxg) [inline src/database/DataTypes/ItemBatch.js:33]: Basically added `breaches` to all the things: from `TransactionItem` to `TransactionBatch` and `StocktakeBatch` `get breaches` is calling `ItemBatch` `get breaches` which is calling `locationMovements get breaches`..

File: src/database/utilities/createRecord.js [hunk 2]
```diff
@@ -786,7 +786,7 @@ const createLocationMovement = (database, itemBatch, location) => {
 };
 
 const createVvmStatusLog = (database, itemBatch, status) => {
-  const vvmStatus = database.create('VaccineVialMonitorStatus', {
+  const vvmStatus = database.create('VaccineVialMonitorStatusLog', {
```
Comment (joshxg) [inline src/database/utilities/createRecord.js:789]: silly bug fix 1

File: src/database/utilities/createRecord.js [hunk 3]
```diff
@@ -943,7 +943,7 @@ export const createRecord = (database, type, ...args) => {
       return createSensor(database, ...args);
     case 'LocationMovement':
       return createLocationMovement(database, ...args);
-    case 'VaccineVialMonitorStatus':
+    case 'VaccineVialMonitorStatusLog':
```
Comment (joshxg) [inline src/database/utilities/createRecord.js:946]: silly bug fix 2

File: src/reducers/BreachReducer.js [hunk 4]
```diff
@@ -20,37 +19,22 @@ export const BreachReducer = (state = initialState(), action) => {
...
-    }
-
-    case BREACH_ACTIONS.SET_FRIDGE_BREACH: {
-      const { payload } = action;
-      const { breachId } = payload;
-
-      const breach = UIDatabase.get('TemperatureBreach', breachId) ?? {};
-
-      const { location: fridge } = breach;
+      const { breaches, fridgeName } = payload;
```
Comment (joshxg) [inline src/reducers/BreachReducer.js:35]: just simplified this reducer so there are three actions: `viewTransactionItemBreaches`, `viewStocktakeBatchBreaches` and `viewFridgeBreaches`, and just the two reducer cases: `viewItemBreaches` and `viewFridgeBreaches` (mainly for setting the title to be "Breaches for fridge xxx" and `"Breaches for Item xx"`....

File: src/selectors/breach.js [hunk 5]
```diff
@@ -11,19 +11,17 @@ export const selectIsBreachModalOpen = ({ breach }) => {
...
-    const { itemName } = batch;
+  if (forBatch || forItem) {
+    const { itemName } = breach;
 
     return `${vaccineStrings.temperature_breaches_for} ${itemName}`;
   }
 
-  const { fridge } = breach;
-  const { description } = fridge;
+  const { fridgeName } = breach;
```
Comment (joshxg) [inline src/selectors/breach.js:24]: I should just make the store have a `name` and then there can reduce a lot of this by a lot, I think

File: src/widgets/modalChildren/BreachDisplay.js [hunk 6]
```diff
@@ -80,7 +81,7 @@ const BreachDisplayComponent = ({ breaches }) => {
 };
 
 BreachDisplayComponent.propTypes = {
-  breaches: PropTypes.object.isRequired,
+  breaches: PropTypes.oneOfType([PropTypes.array, PropTypes.object]).isRequired,
```
Comment (joshxg) [inline src/widgets/modalChildren/BreachDisplay.js:84]: silly bug fix 92

File: src/database/DataTypes/TransactionItem.js [hunk 7]
```diff
@@ -299,6 +299,16 @@ export class TransactionItem extends Realm.Object {
     return this.batches.sum('doses');
   }
 
+  get isInBreach() {
+    return this.batches.some(({ isInBreach }) => isInBreach);
+  }
```
Comment (joshxg) [inline src/database/DataTypes/TransactionItem.js:304]: a `TransactionItem` being essentially a grouping of multiple batches can be in a breach if there is just one of the batches in a breach

File: src/database/DataTypes/TransactionItem.js [hunk 8]
```diff
@@ -299,6 +299,16 @@ export class TransactionItem extends Realm.Object {
...
+
+  get breaches() {
+    return (
+      this.batches?.reduce((acc, { breaches }) => union(acc, breaches, ({ id }) => id), []) ?? []
+    );
```
Comment (joshxg) [inline src/database/DataTypes/TransactionItem.js:309]: essentially taking all distinct `TemperatureBreach` records for the batches

File: src/database/DataTypes/ItemBatch.js [hunk 9]
```diff
@@ -133,22 +137,16 @@ export class ItemBatch extends Realm.Object {
...
    * @param  {Location} newLocation
    * @return {Bool} Indicator whether the new location should be applied to this batch.
    */
-  shouldApplyLocation(newLocation) {
-    const { id: newLocationId } = newLocation;
-    const { id: currentLocationId } = this.location;
-
-    return newLocationId !== currentLocationId;
+  shouldApplyLocation(newLocation = {}) {
+    return newLocation?.id === this.location?.id;
```
Comment (joshxg) [inline src/database/DataTypes/ItemBatch.js:149]: silly bug fix 93 - probs dont need the default

File: src/database/DataTypes/ItemBatch.js [hunk 10]
```diff
@@ -133,22 +137,16 @@ export class ItemBatch extends Realm.Object {
    * @param  {VaccineVialMonitorStatus} newVvmStatus
    * @return {Bool} Indicator whether the new vvm status should be applied to this batch.
    */
-  shouldApplyVvmStatus(newVvmStatus) {
-    const { id: newStatusId } = newVvmStatus;
-    const { id: currentStatusId } = this.currentVvmStatus;
-
-    return newStatusId !== currentStatusId;
+  shouldApplyVvmStatus(newVvmStatus = {}) {
+    return newVvmStatus?.id === this.currentVvmStatus?.id;
```
Comment (joshxg) [inline src/database/DataTypes/ItemBatch.js:141]: silly bug fix 94

File: src/widgets/DataTable/DataTableRow.js [hunk 11]
```diff
@@ -228,6 +229,7 @@ const DataTableRow = React.memo(
                 chevron_right: ChevronRightIcon,
                 history: () => <HistoryIcon color={SUSSOL_ORANGE} />,
                 pencil: () => <PencilIcon color={SUSSOL_ORANGE} />,
+                breach: () => <HazardIcon color={SUSSOL_ORANGE} />,
```
Comment (mark-prins) [inline src/widgets/DataTable/DataTableRow.js:232]: the default color is `SUSSOL_ORANGE` and so doesn't need to be specified. It does however look neater having all three lines look the same :)

File: src/pages/CustomerInvoicePage.js [hunk 12]
```diff
@@ -82,6 +84,8 @@ export const CustomerInvoice = ({
         return onUncheck;
       case 'doses':
         return onEditBatchDoses;
+      case 'isInBreach':
```
[the remark this answers is not in the document]
Comment (joshxg) [reply src/pages/CustomerInvoicePage.js]: nice - I went with `hasBreached` as `someLocation.hasBreached` seemed nicer to me at the time than `someLocation.hasBreach` but now typing this comment I'm not sure again

2752288a09ca8bff9fd4fb88e2c2779b015b58c5

File: src/pages/CustomerInvoicePage.js [hunk 13]
```diff
@@ -64,6 +65,7 @@ export const CustomerInvoice = ({
   onApplyMasterLists,
   route,
   onEditBatchDoses,
+  onViewBreach,
```
[the remark this answers is not in the document]
Comment (joshxg) [reply src/pages/CustomerInvoicePage.js]: yeah for sure!

34818bb91b26809a5ca9aea8df9cd8d7260ef120 and 727e1c3535f390c840b5c534264b396881515d4e

File: src/reducers/BreachReducer.js [hunk 14]
```diff
@@ -3,14 +3,13 @@
  * Sustainable Solutions (NZ) Ltd. 2020
  */
 
-import { UIDatabase } from '../database';
-
 import { BREACH_ACTIONS } from '../actions/BreachActions';
 
 const initialState = () => ({
   isModalOpen: false,
   forBatch: false,
```
[the remark this answers is not in the document]
Comment (joshxg) [reply src/reducers/BreachReducer.js]: amazing 06b61f655f6c8f1b17d0ef776e20f05008320c05

File: src/widgets/DataTable/DataTableRow.js (hunk 11 above)
Comment (joshxg) [reply src/widgets/DataTable/DataTableRow.js]: damnit mark - I couldn't decide what to do. So I removed it! 95fbce7d251693867724fec0e8534afc71f3e41c

File: src/pages/CustomerInvoicePage.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (mark-prins) [reply src/pages/CustomerInvoicePage.js]: I've now said 'breach' to myself too many times and everything sounds weird now. `hasBreached` is fine

**Pull request merged**